Define querier types

ethanfrey opened the pull request:
Begin work on #100 

This is my proposal for a generic Querier interface, with an enum Request we can easily extend (more than adding new methods). I am trying to figure out how to pass rich errors over an FFI boundary (see #199) and will improve the error reporting here once that is merged

@KamiD I would appreciate your review and feedback (as I know you want this feature for a long time and we had some discussions on it)

ethanfrey commented:
Rebased and updated with new ApiError type, I think this is a nicer return value. I'm ready for review.

webmaster128 commented on packages/std/src/types.rs:
```diff
@@ -178,6 +178,29 @@ impl QueryResult {
+    // return value is whatever the contract returns (caller should know)
+    Contract {
+        contract_addr: HumanAddr,
+        msg: Binary, // we pass this in as Vec<u8> to the contract, so allow any binary encoding (later, limit to rawjson?)
```
Couldn't we use `String` as a better JSON encoding? It ensures valid UTF-8, is human readable and does not have the 1/3 encoding overhead of base64.

ethanfrey replied:
A few reasons.

First our Json serializer doesn't escape strings. Thus if you have `"` inside the message body (eg. Any non empty Json message) it will break. This is why I moved to vec  (and then base64). If we really want json with minimal overhead, we would use a raw json type that accepts a valid json object here (not string encoded)

Also, I am still unsure how much we want to commit to json long term for encoding and have minimal hard dependencies (now anything supported by serde) to allow changing it later. We can revisit this decision in 0.8

3rd I wanted to keep consistency between the value passed here to the contract and what the contract receives in the query method. They should be the same type with no processing by the runtime.

webmaster128 replied:
Interesting, thanks

ethanfrey commented:
I would do another pass on it after reflecting over the weekend

ethanfrey commented:
Ok, I am done with the changes. I added a new (Api)SystemError type as recommended by @KamiD for separating out system errors (like no such contract) from errors returned by querying the contract itself.

I also cleaned up (Api)Result handling quite a bit and would look to replace many of the serializable XXXResult types with ApiResult<XXX>, which has nice `Into` and `From` traits.

I also added a `MockQuerier` implementation that allows querying some mock balances right now, but demos using the trait.

ethanfrey commented:
@webmaster128 want to do one more pass here to double-check all the new code?

webmaster128 commented on packages/std/src/api.rs:
```diff
@@ -0,0 +1,349 @@
+            ApiResult::Err(e) => Err(e),
+        }
+    }
+}
```
What is this conversion needed for? Isn't it giving the false impression that `ApiError` and `Error` can be mapped both ways?

ethanfrey replied:
I want to recover normal `Result` types, so all the functionality associated with them is exposed, not just `unwrap()` and `?`, but `map`, `context`, ...

`ApiResult<T, ApiError> => Result<T, ApiError>` just fine. You just can't "upgrade" the error part. This should be defined in the fact that both objects have the same error types.

webmaster128 commented on CHANGELOG.md:
```diff
@@ -41,6 +41,11 @@ a breaking change for the clients.
 - Create explicit `QueryResponse`, analogue to `InitResponse` and `HandleResponse`.
+- Add new `ApiResult` and `ApiError` types to represent serializable counterparts to
+`errors::Result` and `errors::Error`. This has a powerful `Into` and `From` implementations
+to transfer back and forth between `Result` and `ApiResult`.
```
The conversion `Result` -> `ApiResult` is loosing significan information. How can this be a "back and forth" implementation?

ethanfrey replied:
I'll clarify.

ethanfrey replied:
Reworded in https://github.com/CosmWasm/cosmwasm/pull/197/commits/3da90297911cc99c97cf407d96adb9c8bc18aae8
